File: LeetCode_Java/dp/HouseRobberII.java
package dp;

public class HouseRobberII {
    public int rob(int[] nums) {
        if (nums.length == 0 || nums == null){
            return 0;
        }

        if (nums.length == 1) {
            return nums[0];
        }

        int first = robRange(nums, 0, nums.length - 2);
        int second = robRange(nums, 1, nums.length - 1);

        return Math.max(first,second);
    }

    private int robRange(int[] nums, int start, int end) {
        if (start == end) {
            return nums[start];
        }
        int[] dp = new int[nums.length];
        dp[start] = nums[start];
        dp[start + 1] = Math.max(nums[start],nums[start + 1]);

        for (int i = start + 2; i <= end; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i]);
        }

        return dp[end];
    }

    public static void main(String[] args) {
        HouseRobberII solution = new HouseRobberII();

        // Test case
        int[] nums = {2, 3, 2};
        System.out.println(solution.rob(nums)); // Output: 3
    }
}

File: LeetCode_Python/dp/HouseRobberII.py
class HouseRobberII:
    def rob(self, nums: list[int]) -> int:
        def rob_linear(arr: list[int]) -> int:
            n = len(arr)
            if n == 0:
                return 0
            if n == 1:
                return arr[0]
            dp = [0] * n
            dp[0] = arr[0]
            dp[1] = max(arr[0], arr[1])
            for i in range(2, n):
                dp[i] = max(dp[i - 1], dp[i - 2] + arr[i])
            return dp[-1]

        n = len(nums)
        if n == 1:
            return nums[0]

        rob1 = rob_linear(nums[:-1])
        rob2 = rob_linear(nums[1:])

        return max(rob1, rob2)


# Time Complexity: O(n)
# Space Complexity: O(n)

if __name__ == "__main__":
    solution = HouseRobberII()

    print("Expected Output: 3")
    print("Actual Output:", solution.rob([2, 3, 2]))

    print("Expected Output: 4")
    print("Actual Output:", solution.rob([1, 2, 3, 1]))
